### PR TITLE
Use in_guaranteed for let captures

### PR DESCRIPTION
**Pull request by meg-gupta**
With this all let values will be captured with in_guaranteed convention
by the closure. Following are the main changes :

SILGen changes:
- A new CaptureKind::Immutable is introduced, to capture let values as in_guaranteed.
- SILGen of in_guaranteed capture had to be fixed.
  in_guaranteed captures as per convention are consumed by the closure. And so SILGen should not generate a destroy_addr for an in_guaranteed capture.
  But LetValueInitialization can push Dealloc and Release states of the captured arg in the Cleanup stack, and there is no way to access the CleanupHandle and disable the emission of destroy_addr while emitting the captures in SILGenFunction::emitCaptures.
  So we now create, temporary allocation of the in_guaranteed capture iduring SILGenFunction::emitCaptures without emitting destroy_addr for it.

SILOptimizer changes:
- Handle in_guaranteed in CopyForwarding.
- Adjust dealloc_stack of in_guaranteed capture to occur after destroy_addr for on_stack closures in ClosureLifetimeFixup.

IRGen changes :
  - Since HeapLayout can be non-fixed now, make sure emitSize is used conditionally
  - Don't consider ClassPointerSource kind parameter type for fulfillments while generating code for partial apply forwarder.
    The TypeMetadata of ClassPointSource kind sources are not populated in HeapLayout's NecessaryBindings. If we have a generic parameter on the HeapLayout which can be fulfilled by a ClassPointerSource, its TypeMetaData will not be found while constructing the dtor function of the HeapLayout.
    So it is important to skip considering sources of ClassPointerSource kind, so that TypeMetadata of a dependent generic parameters gets populated in HeapLayout's NecessaryBindings.

rdar://58459539

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@meg-gupta before this goes in can you fix up the git history?

**Comment by meg-gupta** [comment]
@gottesmm Will do

**Comment by meg-gupta** [comment]
@swift-ci benchmark

File: lib/IRGen/GenFunc.cpp [hunk 1]
```diff
@@ -625,22 +625,25 @@ static void emitApplyArgument(IRGenFunction &IGF,
                         silConv.getSILType(substParam, substFnTy), in, out);
 }
 
-static CanType getArgumentLoweringType(CanType type,
-                                       SILParameterInfo paramInfo) {
+static CanType getArgumentLoweringType(SILType type,
+                                       SILParameterInfo paramInfo,
+                                       const SILFunction& SILFn,
+				       bool isNoEscape) {
```
Comment (slavapestov) [inline lib/IRGen/GenFunc.cpp]: Nit: please line up with previous line's indent

File: lib/IRGen/GenFunc.cpp [hunk 2]
```diff
@@ -625,22 +625,25 @@ static void emitApplyArgument(IRGenFunction &IGF,
                         silConv.getSILType(substParam, substFnTy), in, out);
 }
 
-static CanType getArgumentLoweringType(CanType type,
-                                       SILParameterInfo paramInfo) {
+static CanType getArgumentLoweringType(SILType type,
+                                       SILParameterInfo paramInfo,
+                                       const SILFunction& SILFn,
```
Comment (slavapestov) [inline lib/IRGen/GenFunc.cpp]: Nit: please put & on the parameter name not the type name

File: lib/IRGen/GenFunc.cpp [hunk 3]
```diff
@@ -625,22 +625,25 @@ static void emitApplyArgument(IRGenFunction &IGF,
...
   case ParameterConvention::Direct_Owned:
   case ParameterConvention::Direct_Unowned:
   case ParameterConvention::Direct_Guaranteed:
   case ParameterConvention::Indirect_In:
   case ParameterConvention::Indirect_In_Constant:
+    return canType;
   case ParameterConvention::Indirect_In_Guaranteed:
-    return type;
-
+    return type.isLoadable(SILFn) || !isNoEscape ? canType : CanInOutType::get(canType);
```
Comment (slavapestov) [inline lib/IRGen/GenFunc.cpp]: This line looks a little bit too long. Maybe use an 'if' instead of ternary operator

File: lib/IRGen/GenFunc.cpp [hunk 4]
```diff
@@ -1057,12 +1061,19 @@ static llvm::Function *emitPartialApplicationForwarder(IRGenModule &IGM,
 
         break;
       }
-      case ParameterConvention::Indirect_In_Guaranteed:
-        // The argument is +0, so we can use the address of the param in
-        // the context directly.
-        param.add(fieldAddr.getAddress());
-        dependsOnContextLifetime = true;
+      case ParameterConvention::Indirect_In_Guaranteed: {
+        if (fieldTy.isLoadable(SILFn) || !outType->isNoEscape()) {
```
Comment (slavapestov) [inline lib/IRGen/GenFunc.cpp]: This is a repeat of the earlier condition. Does it make sense to factor it out?

File: lib/IRGen/GenFunc.cpp [hunk 5]
```diff
@@ -1533,14 +1544,26 @@ Optional<StackAddress> irgen::emitFunctionPartialApplication(
...
-      case ParameterConvention::Indirect_In_Guaranteed: {
+      case ParameterConvention::Indirect_In_Constant:{
         auto addr = fieldLayout.getType().getAddressForPointer(args.claimNext());
         fieldLayout.getType().initializeWithTake(IGF, fieldAddr, addr, fieldTy,
                                                  isOutlined);
         break;
       }
       // Take direct value arguments and inout pointers by value.
+      case ParameterConvention::Indirect_In_Guaranteed: {
+        if (argValTypes[i].isLoadable(SILFn) || !outType->isNoEscape()) {
```
Comment (slavapestov) [inline lib/IRGen/GenFunc.cpp]: Ditto

**Review by slavapestov — COMMENTED** [review]
I'm really happy to see usages of Inout_Aliasable go away.

Ideally we would get rid of `capture.isNoEscape()` too. Is that a possibility? We should be able to promote box captures to stack in the SIL optimizer, instead of relying on Sema/SILGen setting `capture.isNoEscape()`.

**Comment by slavapestov** [comment]
Can we use `inout` instead of `inout_aliasable` for mutable captures as well? If I understand correctly, they already should not alias because of exclusivity checking.

**Comment by gottesmm** [comment]
@slavapestov that is a much larger change. In reality what we need to do is make it so that inout_aliasable is either in_guaranteed or inout. It would let us eliminate a bunch of unfortunate things in the code base.

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@slavapestov In certain cases there is a mismatch between the capture.isNoEscape in the frontend and outType.isNoEscape in IRGen. This can cause issue for generating code for the in_guaranteed param. I am working on changes to get rid of this. Will post an update soon.

**Comment by meg-gupta** [comment]
@swift-ci please test

**Comment by meg-gupta** [comment]
@swift-ci please benchmark

**Comment by meg-gupta** [comment]
@swift-ci please test

**Comment by meg-gupta** [comment]
@swift-ci please benchmark

**Comment by meg-gupta** [comment]
@swift-ci please test

**Comment by meg-gupta** [comment]
@swift-ci Please test OS X platform

**Comment by meg-gupta** [comment]
@swift-ci please benchmark

**Comment by meg-gupta** [comment]
@swift-ci please test

**Comment by meg-gupta** [comment]
@swift-ci Please smoke test OS X platform

**Comment by meg-gupta** [comment]
Onone regressions all seem to come from the temp allocation and copy_addr into it created in SILGenFunction::emitCaptures for in_guaranteed captures.

**Comment by meg-gupta** [comment]
@swift-ci Please Test Source Compatibility

File: lib/IRGen/GenProto.cpp [hunk 6]
```diff
@@ -384,6 +390,10 @@ void PolymorphicConvention::considerParameter(SILParameterInfo param,
     case ParameterConvention::Direct_Owned:
     case ParameterConvention::Direct_Unowned:
     case ParameterConvention::Direct_Guaranteed:
+      // Don't consider ClassPointer source and Metadata source for partial_apply forwader with non fixed layout.
+      // If not, we may end up with missing TypeMetadata for a type dependent generic parameter
+      // while generating code for destructor of HeapLayout.
+      if (!ConsiderParameterSources) return;
```
Comment (aschwaighofer) [inline lib/IRGen/GenProto.cpp]: Is the check here still neccessary if you add a check:  
```
  if (ConsiderParameterSources)
   considerParameter(params[selfIndex], selfIndex, true);
```

to line 221?

Comment (meg-gupta) [reply lib/IRGen/GenProto.cpp]: Forgot to clean it up. Will fix in the following commit.

**Comment by meg-gupta** [comment]
@slavapestov This now handles captures of let values in both escape and non escape closures.

**Comment by meg-gupta** [comment]
@swift-ci please benchmark

**Comment by meg-gupta** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by meg-gupta** [comment]
@swift-ci please test

File: lib/IRGen/GenProto.cpp [hunk 7]
```diff
@@ -217,11 +218,13 @@ PolymorphicConvention::PolymorphicConvention(IRGenModule &IGM,
       considerParameter(params[selfIndex], selfIndex, true);
```
Comment (aschwaighofer) [inline lib/IRGen/GenProto.cpp]: This line needs to be guarded by if (considerParameterSources)

**Comment by meg-gupta** [comment]
@swift-ci please test

**Comment by meg-gupta** [comment]
@swift-ci Please Test Source Compatibility

**Comment by meg-gupta** [comment]
@swift-ci please test

**Comment by meg-gupta** [comment]
@swift-ci Please Test Source Compatibility

**Comment by meg-gupta** [comment]
@swift-ci please test

**Comment by meg-gupta** [comment]
@swift-ci please test

**Comment by meg-gupta** [comment]
@swift-ci please smoke test

**Comment by meg-gupta** [comment]
@swift-ci please smoke test and merge

**Comment by meg-gupta** [comment]
@swift-ci please smoke test

**Comment by atrick** [comment]
> Can we use `inout` instead of `inout_aliasable` for mutable captures as well? If I understand correctly, they already should not alias because of exclusivity checking.

@slavapestov exclusivity is a prerequisite. But now the no-escape closure representation needs to be fixed. `@inout_aliasable` is still used for aliasing stuff

```
func doit(_ f: ()->Int) -> Int {
  f()
}

public func testInoutAlias() -> Int {
  var i = 0
  let f = { i = 3 }
  return doit() {
    f()
    return i
  }
}
```

It's very clear to me how this should be fixed, and I wrote up a design proposal for it years ago but it has been lower priority than bug fixing.

**Review by atrick — COMMENTED** [review]
@meg-gupta I'm very happy you got this working. It's a very clean design that defines away a category of memory bugs.

Please address these issues in a follow-up PR:

- Comments should be 80 columns

In `TypeConverter::getDeclCaptureKind`:

- assert that Immutable captures are address-only

- Fix the above check for opaque values. Note that SILOpaqueValues has
  not effect on SIL calling convention, so whether it is enabled should probably
  have no effect on the CaptureKind.

In TempRValueOpts:

- collectLoads: this looks correct, but please comment why mark_dependence base operand is safe to consider a load, and why the value operand needs to be transitive (otherwise we would miscompile!)

In `irgen::emitFunctionPartialApplication()`:

- In the block comment please explain what is happening clearly. You or @aschwaighofer or @rjmccall  can explain this better than I can, but basically:

> The context's HeapLayout is only "fixed" when all of the captures have
fixed type info. When the HeapLayout is fixed, then [the bindings]
need to map the context's fields to type parameters so that the
context destructor can find the metadata. When the HeapLayout is *not*
fixed, then the metadata is embedded in the context to be read
directly by the destructor, so there is not need to map captures to
type parameters--in this case, bypass computing the bindings because
it would fail when it sees ClassPointer or Metadata sources.

[I could have gotten this completely wrong, even after staring at the code for about an hour, so I think it's well worth explaining]

**Comment by aschwaighofer** [comment]
@atrick The issue is that we *must not* consider parameter sources as type sources (example: we can and do reconstruct type metadata T for a class reference parameter SomeClass<T> in <T> closure_func(@in_guaranteed T, @guaranteed SomeClass<T>)) when computing which metadata to store in the heap layout because -- as opposed to the partially applied function forwarder -- those parameters (the unapplied ones) won't be available in the destructor.

This is only an issue in non fixed heap layouts because only for those we need the metadata of non-fixed captures in the destructor.

**Comment by atrick** [comment]
@aschwaighofer that makes perfect sense, but then why does this PR set `considerParameterSources` to `false` when `HeapLayout` is not fixed. That's what I find so confusing and why it needs to be explained.

**Comment by aschwaighofer** [comment]
"The issue is that we must not consider parameter sources as type sources..."
"This is only an issue in non fixed heap layouts ...."

It is only safe to ``considerParameterSources`` when we know that we won't need type metadata provided by said parameter sources in the destructor (the destructor does not have that parameter source). In oder words: It is only the safe when the HeapLayout is fixed.

There are two sources of type metadata: the ones passed explictitly to a generic function (or stored explicitly in the heap context), and those implicit in other parameters (e.g SomeClass<T> parameter). We must ignore the SomeClass<T> parameter source because it is not available in the destructor but only in the partial apply forwarder.

All this is only an issue for non-fixed types because for fixed ones we don't need that Metadata.

**Comment by gottesmm** [comment]
@aschwaighofer add that in a comment? Just doing a drive by.

**Comment by meg-gupta** [comment]
@gottesmm I'll include it in a follow up PR with other suggestions above

**Comment by atrick** [comment]
@aschwaighofer thanks

My confusion had to do with what the "bindings" for type metadata are
used for... Let me try again:

```
The context's HeapLayout is "fixed" when all of its captures [partially
applied parameters?] have a fixed layout. A context with fixed
HeapLayout can be destroyed without access to type metadata;
consequently, it is safe to consider partially applied parameters as
the source of type metadata bindings even though those parameters are
unavailable when the context is destroyed.

When the context's HeapLayout is not fixed, then its destructor
requires access to type metadata of each capture. Therefore, partially
parameters, which are unavailable when the destructor is called, cannot
be the souce of type metadata.
```

(I'm not actually sure what the comment about ClassPointer or Metadata
sources has to do with, presumably there are other sources?)
